### PR TITLE
[M44917]: fixed misalignment of ... menu

### PR DESCRIPTION
**Pull request by michelengelen**
#### Summary
the CSSTransition was messing with the calculations made by popperJS and since there is no easy way to automatically update that I implemented floatingUI instead which is far easier to use and also has a built in autoUpdate function that re-calculates the position on every change to the component.

#### Ticket Link
[MM-44917](https://mattermost.atlassian.net/browse/MM-44917)

#### Related Pull Requests
n/a

#### Screenshots
n/a

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by AshishDhama** [comment]
@michelengelen Does it make sense to migrate popper to FloatingUI, I think there aren’t many use-cases, BTW Floating UI doesn’t have a major version yet.

**Comment by michelengelen** [comment]
> @michelengelen Does it make sense to migrate popper to FloatingUI, I think there aren’t many use-cases, BTW Floating UI doesn’t have a major version yet.

IMO it is floatingUI better than popperJS ... the fact that it does not have a major version yet should not hold us back from adding it in. The calculations and the ability to auto update them are main reasons I would almost always go with floatingUI.

The fact that we now have both is a bit of a downside to it (although the impact is really small), but once we have compass components we should evaluate the usage of external UI libraries anyways.

**Review by AshishDhama — APPROVED** [review]
LGTM!

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Looks good @michelengelen. One small change: can we bump the popover down so that it is 4px from the top of the ••• button? Just so it feels more connected to the button.

**Review by matthewbirtch — APPROVED** [review]

**Comment by michelengelen** [comment]
FYI: @AshishDhama I fixed the Javascript error, since the `autoUpdate` function uses a `ResizeObserver` as well and that somehow messes with the one used in the `useFormattingBarControls` custom hook. The error itself was not harmful, but it still made the announcement bar show up

File: components/advanced_text_editor/formatting_bar/formatting_bar.tsx [hunk 1]
```diff
@@ -285,14 +277,17 @@ const FormattingBar = (props: FormattingBarProps): JSX.Element => {
...
             <HiddenControlsContainer
-                ref={popperRef}
-                style={{...popper, zIndex: 2}}
-                {...attributes.popper}
+                ref={floating}
+                style={{
+                    position: strategy,
+                    top: y ?? 0,
+                    left: x ?? 0,
+                    zIndex: 20,
```
Comment (koox00) [inline components/advanced_text_editor/formatting_bar/formatting_bar.tsx:285]: using an object literal here will make `HiddenControlsContainer` re-render on every parent re-render.
Dunno though if it makes sense to `useMemo` the object or not.

**Review by koox00 — APPROVED** [review]

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
Expand and collapse issue is fixed. The popover menu is still disconnected from 3dots on longer posts.
Screenshot 2022-06-09 095600

**Comment by jgilliam17** [comment]
Thanks @michelengelen the floating popover menu is fixed
[E2Es](https://mattermost-cypress-report.s3.amazonaws.com/205301-1686201-onprem-ent-webapp-pr-10529-mattermost/mm-ee-test1686201/mochawesome.html) show 20 failures related to `HiddenControlsContainer` in addition to other tests that are failing related to text editor that will be fixed in our other follow up ticket. Can you please take a look and fix the HiddenControl related ones here in this PR? Let me know - thanks

**Comment by michelengelen** [comment]
@jgilliam17 could you list the tests that are failing? There are 87 and I am not sure which ones are related to this PR

**Comment by michelengelen** [comment]
@jgilliam I think with this fix the failing tests should now be fixed ... I will re-run E2Es now. Sorry for all the noise!

**Review by hmhealey — DISMISSED** [review]
Is floating-ui strictly necessary for this? We need to get everyone on the same page about these various tooltip libraries since this would be the 4th one we have in the web app now.

**Comment by michelengelen** [comment]
> Is floating-ui strictly necessary for this? We need to get everyone on the same page about these various tooltip libraries since this would be the 4th one we have in the web app now.

it is not strictly necessary, but popperJS had some issues with correctly placing the overlay. I am 4/5 that floatingUI will resolve some of the issues we have with the other libraries, since it is built way better with more functionality baked in and the ability to extend it using custom middlewares.

As a second point the whole package has only 337KB (consisting of `@floating-ui/core`, `@floating-ui/dom` and `@floating-ui/react-dom`) as opposed to `popper`s ~2MB (`@popper-js/core` at ~1.7MB, `react-popper` at 100KB and `react-popper-tooltip` at ~112KB)

Last but not least (massively relevant if you ask me) popperJS is a discontinued package, so sticking with it will hurt us in the long run.

**Comment by AshishDhama** [comment]
@michelengelen @hmhealey I did a global search and found only two use cases of popper.js. I think we can create Community tickets for both of them. About using the others, I know of one `tippy`. My original plan was to remove tippy and use Floating UI, waiting for Christopher's `common components` effort to finalise. Will migrate `tour tip` to `common components` with necessary changes to remove `tippy`
Screenshot 2022-06-14 at 3 45 11 PM

**Comment by hmhealey** [comment]
I don't necessarily think using old libraries is that bad as long as they're not actively causing us problems, but I do agree that Floating UI is probably our best bet for moving forward. I want to make sure though that we actually address the issue of duplicated dependencies rather than compounding the problem. It doesn't particularly help if Floating UI is smaller as long as we're still also using Popper.

I'd be fine with having both libraries for now as long as we have a concrete plan to follow up with removing Popper in the next week or two though. We can't keep letting these dependencies pile up, and I'm worried that leaving it to the community or until the common components are eventually done would just mean it gets forgotten about as the web app keeps increasing in size.

And for the other ones, I was thinking of Tippy and React Bootstrap. Obviously, React Bootstrap is a mess of tech debt that we'll have to deal with on its own though

**Comment by michelengelen** [comment]
> I don't necessarily think using old libraries is that bad as long as they're not actively causing us problems, but I do agree that Floating UI is probably our best bet for moving forward. I want to make sure though that we actually address the issue of duplicated dependencies rather than compounding the problem. It doesn't particularly help if Floating UI is smaller as long as we're still also using Popper.
> 
> I'd be fine with having both libraries for now as long as we have a concrete plan to follow up with removing Popper in the next week or two though. We can't keep letting these dependencies pile up, and I'm worried that leaving it to the community or until the common components are eventually done would just mean it gets forgotten about as the web app keeps increasing in size.
> 
> And for the other ones, I was thinking of Tippy and React Bootstrap. Obviously, React Bootstrap is a mess of tech debt that we'll have to deal with on its own though

I agree that we should not let that pile up. I will create a follow-up ticket to replace our immediate usages with floatingUI. One thing I just noticed is that tippy uses popper under the hood as well. so until we remove the tippy usages we will not 100% get rid of popper for now, but at least we are not using it directly anymore

**Comment by hmhealey** [comment]
That's really good to know. I figured they had their own version of that same logic as well.

Thanks for creating that followup. I'll unblock this then since we're all on the same page about converging on Floating UI eventually for this sort of thing

**Review by jgilliam17 — APPROVED** [review]
Thanks @michelengelen 
Tested, looks good to merge.
Latest [E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/208516-945ad3a-onprem-ent-webapp-pr-10529-mattermost/mm-ee-test945ad3a/mochawesome.html) for reference, failures are known, addressed separately.

**Pull request merged**